--- libdrm-2.4.99/util_double_list.h.orig	2019-07-02 20:36:50.000000000 +0000
+++ libdrm-2.4.99/util_double_list.h	2019-08-16 15:50:46.133276669 +0000
@@ -110,7 +110,7 @@
 #ifndef container_of
 #define container_of(ptr, sample, member)				\
     (void *)((char *)(ptr)						\
-	     - ((char *)&((typeof(sample))0)->member))
+	     - ((char *)&((__typeof(sample))0)->member))
 #endif
 
 #define LIST_FOR_EACH_ENTRY(pos, head, member)				\
