--- libdrm-2.4.99/xf86drm.h.orig	2019-07-02 20:36:50.000000000 +0000
+++ libdrm-2.4.99/xf86drm.h	2019-08-16 16:28:37.995074877 +0000
@@ -472,6 +472,17 @@
 #endif /* architecture */
 #endif /* __GNUC__ >= 2 */
 
+#if defined(__SUNPRO_C)
+#include <atomic.h>
+#define atomic_cmpset_int(p, c, n) ((c == atomic_cas_uint(p, c, n)) ? 1 : 0)
+#define DRM_CAS(lock,old,new,__ret)          \
+               do {                          \
+                                       unsigned int __result, __old = (old);\
+                                       __result = !atomic_cmpset_int(lock,__old,new);\
+                                       __ret = __result;          \
+                               } while(0)
+#endif
+
 #ifndef DRM_CAS
 #define DRM_CAS(lock,old,new,ret) do { ret=1; } while (0) /* FAST LOCK FAILS */
 #endif
@@ -889,6 +900,14 @@
 			      uint32_t src_handle, uint64_t src_point,
 			      uint32_t flags);
 
+#ifdef __sun
+extern int _sun_drm_get_major(void);
+extern int _sun_drm_find_device(int min, char **pathp);
+extern int _sun_drm_get_subsystem(char *path);
+extern int _sun_drm_get_pci_bus_info(char *path, drmPciBusInfo *bus);
+extern int _sun_drm_get_pci_dev_info(char *path, drmPciDeviceInfo *dev);
+#endif /* __sun */
+
 #if defined(__cplusplus)
 }
 #endif
